src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingExtensionMethods.cs

```diff
@@ -0,0 +1,132 @@
+
+            IEnumerable<DatabaseObjectType> objectTypes = publishModel.GetDatabaseObjectTypes();
+            foreach (DatabaseObjectType objectType in objectTypes)
+            {
+                IEnumerable<KeyValuePair<string, string>> databaseObjectsOfType = publishModel.EnumChildrenForDatabaseObjectType(objectType);
+
+                Logger.Write(
+                    LogLevel.Normal,
+                    string.Format(
+                        "Loaded SMO urn object count {0} for type {1}, urns: {2}",
+                        objectType,
+                        databaseObjectsOfType.Count(),
+                        string.Join(", ", databaseObjectsOfType.Select(p => p.Value))));
+
+                foreach (KeyValuePair<string, string> databaseObjectOfType in databaseObjectsOfType)
+                {
+                    if (!serverAndDatabaseInitialized)
+                    {
+                        Urn urn = new Urn(databaseObjectOfType.Value);
+                        serverName = urn.GetNameForType("Server");
+                        databaseName = urn.GetNameForType("Database");
+                        serverAndDatabaseInitialized = true;
+                    }
+
+                    databaseObjects.Add(new Urn(databaseObjectOfType.Value).ToScriptingObject());
+                }
+            }
+
+            return databaseObjects;
+        }
+
+        public static bool IsOperationCanceledException(this Exception e)
+        {
+            Exception current = e;
+            while (current != null)
+            {
+                if (current is OperationCanceledException)
+                {
+                    return true;
+                }
+
+                current = current.InnerException;
+            }
+
+            return false;
+        }
+
+        public static Urn ToUrn(this ScriptingObject scriptingObject, string server, string database)
+        {
+            if (scriptingObject == null)
+            {
+                throw new ArgumentNullException("scriptingObjectc");
+            }
+
+            if (string.IsNullOrWhiteSpace(server))
+            {
+                throw new ArgumentException("Parameter must have a value", "server");
+            }
+
+            if (string.IsNullOrWhiteSpace(database))
+            {
+                throw new ArgumentException("Parameter must have a value", "database");
+            }
+
+            if (string.IsNullOrWhiteSpace(scriptingObject.Name))
+            {
+                throw new ArgumentException("Property scriptingObject.Name must have a value", "scriptingObject");
+            }
+
+            if (string.IsNullOrWhiteSpace(scriptingObject.Type))
+            {
+                throw new ArgumentException("Property scriptingObject.Type must have a value", "scriptingObject");
+            }
+
+            string urn = string.Format(
+                "Server[@Name='{0}']/Database[@Name='{1}']/{2}[@Name='{3}' {4}]",
```

Comment:
>Server[@Name='{0}']/Database[@Name='{1}']/{2}[@Name='{3}' {4}]", [](start = 17, length = 64)

@llali - it's worth looking at this for ObjectExplorer. This is how you use URN instead of the direct SMO object model. We should evaluate this in the future for PERF reasons, particularly as we implement filtering since this has good filter support. I'd be interested to see if this has significantly better perf in the long-run vs. using the object model. Note that Brian has done the hard work of adding InternalsVisibleTo for our project to the SMO DLLs that needed updating :)